clippy_lints/src/lifetimes_bound_nested_ref.rs

```diff
@@ -0,0 +1,222 @@
+/// Lints to help dealing with unsoundness due to a compiler bug described here:
+///
+/// https://github.com/rust-lang/rustc-dev-guide/blob/478a77a902f64e5128e7164e4e8a3980cfe4b133/src/traits/implied-bounds.md .
+///
+/// For the following three cases the current compiler (1.76.0) gives a later error message when
+/// manually adding a generic lifetime bound that is implied by a nested reference:
+///
+///     https://github.com/rust-lang/rust/issues/25860
+///     Implied bounds on nested references + variance = soundness hole
+///     https://github.com/rust-lang/rust/issues/25860#issue-82044592
+///     
+///     https://github.com/rust-lang/rust/issues/84591
+///     HRTB on subtrait unsoundly provides HTRB on supertrait with weaker implied bounds
+///     
+///     https://github.com/rust-lang/rust/issues/100051
+///     implied bounds from projections in impl header can be unsound
+///     
+/// The lints here suggest to manually add such lifetime bounds in the hope that
+/// the unsoundness is avoided.
+///
+/// There are also reverse lints that suggest to remove lifetime bounds
+/// that are implied by nested references. These lints are intended to be used only
+/// after the compiler handles these lifetime bounds correctly.
+///
+/// All lints here are in the nursery category.
+use clippy_utils::diagnostics::span_lint;
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::{GenericBound, Lifetime, Ty, TyKind, WherePredicate};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::impl_lint_pass;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks function arguments and return values that have a nested reference type with lifetimes,
+    /// and suggests to add the implied generic lifetime bounds.
+    /// Adding a lifetimes bound helps to avoid unsound code because this addition
+    /// can lead to a compiler error in related source code, as observed in rustc 1.76.0.
+    ///
+    /// ### Why is this bad?
+    /// This is described in: https://github.com/rust-lang/rust/issues/25860
+    /// and as one case of unsoundness here:
+    /// https://github.com/rust-lang/rustc-dev-guide/blob/478a77a902f64e5128e7164e4e8a3980cfe4b133/src/traits/implied-bounds.md .
+    ///
+    /// ### Known problems
+    /// It is not known whether this covers all cases in issue 25860.
+    ///
+    /// ### Example, the val_a argument implies a lifetimes bound:
+    /// ```no_run
+    /// pub const fn lifetime_translator<'a, 'b, T>(val_a: &'a &'b (), val_b: &'b T) -> &'a T {...}
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// pub const fn lifetime_translator<'a, 'b: 'a, T>(val_a: &'a &'b (), val_b: &'b T) -> &'a T {...}
+    /// ```
+    #[clippy::version = "1.78.0"]
+    pub ADD_REDUNDANT_LIFETIMES_BOUND_NESTED_REF_FN,
```

Comment:
Those lint names are a mouthful.

Maybe: `IMPLICIT_LIFETIME_BOUNDS` here and `EXPLICIT_*` for the lint that removes those again?

Reply:
> Those lint names are a mouthful.
> 
> Maybe: `IMPLICIT_LIFETIME_BOUNDS` here and `EXPLICIT_*` for the lint that removes those again?

From back to front:

The ```_FN``` suffix is there to indicate to the 25860 issue. The idea was to start of with different suffixes for the two issues but merging the lints into the same name makes them easier to use, so I'll drop the suffix.

```LIFETIME_BOUNDS``` is easier to read by not accurate: the lint gives an error message for a single bound, 
and each bound has two lifetimes.

I'll change the lint names to use ```IMPLICIT``` instead of ```ADD_REDUNDANT```, and similarly for the explicit and remove redundant.

Reply:
> `LIFETIME_BOUNDS` is easier to read by not accurate

Please refer to the [lint naming conventions](https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#lints), 2. paragraph:

> The basic rule is: the lint name should make sense when read as "allow lint-name" or "allow lint-name items". For example, "allow deprecated items" and "allow dead_code" makes sense, while "allow unsafe_block" is ungrammatical (should be plural).

---

I think you can keep the `_FN` suffix to separate those lints, but I'm not convinced that the `NESTED_REF` really has to be part of the lint name.